iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp

```diff
@@ -0,0 +1,278 @@
+struct HasValueMethod<T, cxx::void_t<decltype(std::declval<T>().value())>> : std::true_type
+{
+};
+
+template <typename T, class = void>
+struct HasGetErrorMethod : std::false_type
+{
+};
+
+template <typename T>
+struct HasGetErrorMethod<T, cxx::void_t<decltype(std::declval<T>().get_error())>> : std::true_type
+{
+};
+
+template <typename T>
+struct Expect
+{
+    /// @brief Expects that the object is valid, otherwise the method prints the
+    ///        provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
+    void expect(const char* const msg) const noexcept;
+};
+
+template <typename T, typename ValueType>
+struct ExpectWithValue
+{
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
+    /// @return the contained value type
+    ValueType& expect(const char* const msg) & noexcept;
+
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
+    /// @return the contained value type
+    const ValueType& expect(const char* const msg) const& noexcept;
+
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
+    /// @return the contained value type
+    ValueType&& expect(const char* const msg) && noexcept;
+
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be print when the object is invalid
+    /// @return the contained value type
+    const ValueType&& expect(const char* const msg) const&& noexcept;
+};
+
+template <typename T, typename ValueType>
+struct ValueOr
+{
+    /// @brief When the object contains a value a copy will be returned otherwise a
+    ///        a copy of alternative will be returned
+    /// @param[in]  alternative the return value which will be used when the object does
+    ///             not contain a value
+    /// @return A copy of the contained value if possible otherwise a copy of alternative
+    ValueType value_or(const ValueType& alternative) const noexcept;
+};
+
+template <typename T, typename ValueType>
+struct AndThenWithValue
+{
+    using and_then_callback_t = cxx::function_ref<void(ValueType&)>;
+    using const_and_then_callback_t = cxx::function_ref<void(const ValueType&)>;
+
+    /// @brief Calls the provided callable when the object is valid and provide the underlying
```

Comment:
```suggestion
    /// @brief Calls the provided callable when the object is valid and provides the underlying
```
Please correct this also in the following comments.